// ros2nodejs/lib/Action.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Action = void 0;
class Action {
    constructor(ros, name, actionType) {
        this.ros = ros;
        this.name = name;
        this.actionType = actionType;
    }
    sendGoal(goal, successCallback, failedCallback, feedbackCallback) {
        const ID = this.ros.getRandomUUID();
        this.ros.on(ID, (message) => {
            if (message.response_type === 'feedback') {
                feedbackCallback === null || feedbackCallback === void 0 ? void 0 : feedbackCallback(message.values);
            }
            else if (message.response_type === 'result') {
                successCallback === null || successCallback === void 0 ? void 0 : successCallback(message.values);
            }
            else {
                failedCallback === null || failedCallback === void 0 ? void 0 : failedCallback(message.values);
            }
        });
        this.ros.send({
            op: 'send_goal',
            id: ID,
            action_name: this.name,
            action_type: this.actionType,
            feedback: true,
            goal_msg: goal,
        });
    }
    destroyClient() {
        this.ros.send({
            op: 'destroy_client',
            action_type: this.actionType,
        });
    }
    cancelGoal(callback) {
        const ID = this.ros.getRandomUUID();
        this.ros.on(ID, (message) => {
            callback === null || callback === void 0 ? void 0 : callback(message.values);
        });
        this.ros.send({
            op: 'cancel_goal',
            id: ID,
            action_type: this.actionType,
        });
    }
}
exports.Action = Action;
